clippy_lints/src/redundant_guard.rs

```diff
@@ -0,0 +1,277 @@
+                    expr,
+                    outer_arm,
+                    if_expr.span,
+                    pat_binding,
+                    pat.span,
+                    None,
+                    &mut app,
+                );
+
+                continue;
+            }
+        }
+    }
+}
+
+fn remove_binding(
+    cx: &LateContext<'_>,
+    pat: &Pat<'_>,
+    span: Span,
+    app: &mut Applicability,
+) -> (Cow<'static, str>, Cow<'static, str>) {
+    let pat_start = snippet_with_applicability(cx, pat.span.with_hi(span.lo()), "<start>", app);
+    let pat_end = snippet_with_applicability(cx, pat.span.with_lo(span.hi()), "<end>", app);
+    (pat_start, pat_end)
+}
+
+fn get_pat_binding(cx: &LateContext<'_>, guard_expr: &Expr<'_>, outer_arm: &Arm<'_>) -> Option<Span> {
+    if let ExprKind::Path(qpath) = guard_expr.kind
+        && let res = cx.qpath_res(&qpath, guard_expr.hir_id)
+        && local_came_from(cx, &res, outer_arm.pat.hir_id)
+    {
+        return cx.tcx.hir().res_span(res);
+    }
+
+    None
+}
+
+fn get_guard(cx: &LateContext<'_>, guard: Option<Guard<'_>>, app: &mut Applicability) -> Cow<'static, str> {
+    guard.map_or_else(
+        || Cow::Borrowed(""),
+        |guard| {
+            let (prefix, span) = match guard {
+                Guard::If(e) => ("if", e.span),
+                Guard::IfLet(l) => ("if let", l.span),
+            };
+
+            Cow::Owned(format!(
+                " {prefix} {}",
+                snippet_with_applicability(cx, span, "<guard>", app),
+            ))
+        },
+    )
+}
+
+fn local_came_from(cx: &LateContext<'_>, local: &Res, target: HirId) -> bool {
+    if let Res::Local(local) = local && cx.tcx.hir().parent_iter(*local).any(|(p, _)| p == target) {
+        return true;
+    }
+
+    false
+}
+
+#[expect(clippy::too_many_arguments)]
+fn emit_redundant_guard<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &Expr<'tcx>,
+    outer_arm: &Arm<'_>,
+    guard_span: Span,
+    pat_binding: Span,
+    pat_span: Span,
+    inner_guard: Option<Guard<'_>>,
+    app: &mut Applicability,
+) {
+    if is_from_proc_macro(cx, expr) {
+        return;
+    }
+
+    let (pat_start, pat_end) = remove_binding(cx, outer_arm.pat, pat_binding, app);
+    let binding_replacement = snippet_with_applicability(cx, pat_span, "<binding_repl>", app);
+
+    span_lint_and_then(
+        cx,
+        REDUNDANT_GUARD,
+        guard_span.source_callsite(),
+        "redundant guard",
+        |diag| {
+            diag.span_suggestion(
+                outer_arm.span.with_hi(guard_span.source_callsite().hi()),
+                "try",
+                format!(
+                    "{pat_start}{binding_replacement}{pat_end}{}",
+                    get_guard(cx, inner_guard, app),
+                ),
+                *app,
+            );
+        },
+    );
+}
+
+/// Checks if the given `Expr` can also be represented as a `Pat`.
+fn expr_can_be_pat(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
```

Comment:
I've been looking at this for the past hour. Maybe I need some help...
I think it's in line 233 or 263 though :D

Reply:
I'm confused too, where is it?

Reply:
The one I had in mind was where the pattern was syntactically valid but wouldn't be applicable to the type it's matching on, as an example:

```rust
let v = Some(vec![1u8, 2, 3]);
match v {
    Some(x) if x == [1] => {},
    _ => {},
}
```

As much as it would be nice the compiler doesn't accept `Some([1])` as a pattern for an `Option<Vec<T>>`

Reply:
> The one I had in mind was where the pattern was syntactically valid but wouldn't be applicable to the type it's matching on, as an example:

Yeah that could be an issue, though the compiler suggests `v.as_deref()` in this case plus it's `MaybeIncorrect`. Is that fine enough?

Reply:
Though the rest pattern should definitely be fixed

Reply:
I think it would have too many FPs personally, the `Option<Vec<>>` case ain't too bad as you say but there's things like

```rust
fn f(s: Option<std::ffi::OsString>) {
    match s {
        Some(x) if x == "a" => {}
        _ => {}
    }
}
```

Or where the outer type isn't an `Option` that has a nice conversion method

Reply:
We could always ignore `if x ==` then. `if let` and `if matches!` must be patterns already, anyway.

Reply:
If we can keep `==` that would be great since I expect this to be fairly common, but if it's too much trouble yeah it could be left as a follow up to be done

Reply:
Yeah that should do it 👍